lsp/nls/src/analysis.rs

```diff
@@ -549,230 +545,243 @@ impl PackedAnalysis {
+            let initial_type_ctxt =
+                mk_initial_ctxt(alloc, std::iter::once((StdlibModule::Std, *slf.ast)))
+                    .expect("fail to create the initial typing environment");
 
-        // Safety: everything that we store in the current analysis is borrowed from `self.alloc`,
-        // which is guaranteed to live as long as `self`.
-        self.analysis = unsafe {
-            std::mem::transmute::<Analysis<'ast>, Analysis<'static>>(Analysis::new(
+            let type_tables = typecheck_visit(
                 alloc,
-                ast,
-                type_lookups,
-                &Environment::new(),
-            ))
-        };
+                slf.ast,
+                initial_type_ctxt.clone(),
+                &mut StdlibResolver,
+                &mut collector,
+                TypecheckMode::Walk,
+            )
+            .expect("failed to typecheck `std`");
 
-        initial_type_ctxt
-    }
-}
+            let type_lookups = collector.complete(alloc, type_tables);
 
-/// Re-borrow the packed analysis' falsely `static` AST with a lifetime tied to its allocator. This
-/// is an internal function, you should use the macro [borrow_ast!] instead.
-///
-/// ## Soft Safety
-///
-/// The following is not part of the safety contract per se for this free-standing function, but is
-/// a safety invariant to maintain when this function is used in the context of [PackedAnalysis].
-///
-/// The caller must ensure that the asts stored in `asts` have been allocated with `_alloc` (or any
-/// allocator that will live as long as `_alloc`).
-fn borrow_ast<'ast>(_alloc: &'ast AstAlloc, ast: &'ast Ast<'static>) -> &'ast Ast<'ast> {
-    ast
-}
+            *slf.type_ctxt = Some(initial_type_ctxt);
+            *slf.analysis =
+                Analysis::new(alloc, slf.ast, type_lookups, &Environment::<'static>::new());
+        })
+    }
 
-/// Same as [borrow_ast()], but for the analysis.
-///
-/// # Safety
-///
-/// The caller must ensure that any data inserted in the analysis must've been allocated with
-/// `_alloc`, or at least be guaranteed to live as long as `_alloc`.
-///
-/// ## Soft Safety
-///
-/// The following is not part of the safety contract per se for this free-standing function, but is
-/// a safety invariant to maintain when this function is used in the context of [PackedAnalysis].
-///
-/// The caller must ensure that the data stored in `analysis` have been allocated with `_alloc` (or
-/// any allocator that will live as long as `_alloc`).
-unsafe fn borrow_analysis_mut<'a, 'ast>(
-    _alloc: &'ast AstAlloc,
-    analysis: &'a mut Analysis<'static>,
-) -> &'a mut Analysis<'ast> where {
-    std::mem::transmute::<&'a mut Analysis<'static>, &'a mut Analysis<'ast>>(analysis)
+    /// If this is the packed analysis of the stdlib, returns its type context.
+    ///
+    /// If this is any other packed analysis, panics.
+    pub(crate) fn stdlib_type_context(&self) -> &TypeContext {
```

Comment:
There's probably a good reason for it, but the specification of this method sounds slightly odd. Is there no simpler way to get the stdlib context from the whole registry, first fetching the stdlib analysis and then its type context without being at risk of panicking?